arch/rv32i/src/lib.rs

```diff
@@ -0,0 +1,176 @@
+#![crate_name = "rv32i"]
+#![crate_type = "rlib"]
+#![feature(asm, const_fn, lang_items, global_asm)]
+#![feature(crate_visibility_modifier)]
+#![no_std]
+
+pub mod machine_timer;
+pub mod plic;
+pub mod support;
+
+extern "C" {
+    // Where the end of the stack region is (and hence where the stack should
+    // start).
+    static _estack: u32;
+
+    // Address of _start_trap.
+    static _start_trap: u32;
+
+    // Boundaries of the .bss section.
+    static mut _szero: u32;
+    static mut _ezero: u32;
+
+    // Where the .data section is stored in flash.
+    static mut _etext: u32;
+
+    // Boundaries of the .data section.
+    static mut _srelocate: u32;
+    static mut _erelocate: u32;
+}
+
+// Entry point of all programs (_start).
+//
+// It initializes DWARF call frame information, the stack pointer, the
+// frame pointer (needed for closures to work in start_rust) and the global
+// pointer. Then it calls _start_rust.
+#[cfg(any(target_arch = "riscv32", target_arch = "riscv64"))]
+global_asm!(
```

Comment:
A while ago we talked about migrating away from inline assembly and instead just relying on assembly, linked as C functions. IIRC this was because of the optimization tricks the compiler does such that the generated assembly may (unless you get all of the annotations right) not be what you wrote.

Reply:
Ok, yes, this will be fixed in the next PR. It's fixed in the `riscv2` branch.